crates/types/src/traits/block_contents.rs

```diff
@@ -64,8 +70,12 @@ pub trait BlockHeader:
     /// Block payload associated with the commitment.
     type Payload: BlockPayload;
 
-    /// Build a header with the payload commitment and parent header.
-    fn new(payload_commitment: Commitment<Self::Payload>, parent_header: &Self) -> Self;
+    /// Build a header with the payload commitment, metadata, and parent header.
```

Comment:
It's not clear to me that `BlockHeader` needs to be a trait. Will the sequencer impl this trait? Does the sequencer even know anything about block headers?

Maybe it should be a struct like
```rust
struct BlockHeader<METADATA> {
  payload_commitment: VID::Commitment, // ...or something
  metadata: METADATA,
  parent: &Self,
}
```

Reply:
btw @jbearer what will be in `Metadata` other than the namespace table? If we put the namespace table inside the payload like we've been discussing then perhaps `Metadata` will just be `()` in the sequencer. Seems kinda crufty to keep this `Metadata` around if we don't actually need it.

Reply:
The header is extremely important to the sequencer. It's where we put all the data that gets broadcast to replicas and validated. E.g. fee payments, L1 info, timestamps, etc....

If we put the namespace table in the payload, `Metadata` will be something like `u32`, not `()`. Because we need the size of the namespace table in the block header as a starting point for interpreting the payload.

Reply:
I guess we could define the size as the first 4 bytes of payload, and that would still give a unique interpretation. But we can save a lookup/evaluation proof for 4 bytes in the header, which still might be useful

Reply:
Ok so I think, what we basically have with the transactions table idea, is a canonical way to encode a generic list of variably sized elements, with (somewhat) efficient inclusion proofs.

Why don't we do the exact same thing with the namespace table in the beginning of the payload, where the namespace table is a list of (offset, size, rollup1, rollup2, ....) elements. This lets us do transaction inclusion proofs by combining a list inclusion proof for the namespace table entry and a list inclusion proof for the transaction data. In this case we don't care about finding all namespaces that contain a particular rollup, we just need the prover to point us at the one particular namespace that includes our transaction.

For namespace proofs, ie rollup full nodes, we will have the verifier download the entire namespace table, which we had them doing before anyways. This lets them find all namespaces relevant to their rollup, and check that each bundle is included in every applicable rollup. The only extra cost is the computation of one additional KZG commitment -- assuming the namespace table fits within a single polynomial, which seems a reasonable enough assumption.

But a malicious builder could really screw over a rollup by making the namespace table huge and forcing the rollup to do a bunch of extra KZG commitments even if they don't have any transactions in the block. I think this is a decent reason to put the namespace table size in the block header as `Metadata`: by _defining_ the namespace table as the first `size` bytes of payload, with `size` coming from the header, we can enforce a maximum size or fee per size in consensus.